### PR TITLE
AK-22097 Ath on AWS CXPs

### PR DESCRIPTION
**Pull request by asim-ak**
added peering_gateway_cxp resource and data source.
updated azure_vnet_connector to support cxpPeeringGatewayId

### CODE REVIEW

File: alkira/resource_alkira_connector_azure_vnet.go [hunk 1]
```diff
@@ -64,7 +64,8 @@ func resourceAlkiraConnectorAzureVnet() *schema.Resource {
 				Optional: true,
 				Default:  "VNET_GATEWAY",
 				ValidateFunc: validation.StringInSlice([]string{
-					"VNET_GATEWAY", "VNET_PEERING"}, false),
+					"VNET_GATEWAY", "VNET_PEERING",
+				}, false),
```
Comment (spikewang) [inline alkira/resource_alkira_connector_azure_vnet.go:68]: what's changed here except new line?

File: alkira/resource_alkira_connector_azure_vnet.go [hunk 2]
```diff
@@ -109,7 +110,8 @@ func resourceAlkiraConnectorAzureVnet() *schema.Resource {
 				Default:  "ADVERTISE_DEFAULT_ROUTE",
 				ValidateFunc: validation.StringInSlice([]string{
 					"ADVERTISE_DEFAULT_ROUTE",
-					"ADVERTISE_CUSTOM_PREFIX"}, false),
+					"ADVERTISE_CUSTOM_PREFIX",
+				}, false),
```
Comment (spikewang) [inline alkira/resource_alkira_connector_azure_vnet.go:114]: same here, this is not really related to what this PR should do?

File: alkira/resource_alkira_connector_azure_vnet.go [hunk 3]
```diff
@@ -243,18 +248,21 @@ func resourceAlkiraConnectorAzureVnet() *schema.Resource {
 				Type:        schema.TypeString,
 				Optional:    true,
 			},
+			"cxp_peering_gateway_id": {
```
Comment (spikewang) [inline alkira/resource_alkira_connector_azure_vnet.go:251]: I think this field should be consistent with resource name, which is `peering_gateway_cxp_id`

File: alkira/resource_alkira_connector_azure_vnet.go [hunk 4]
```diff
@@ -403,12 +406,10 @@ func resourceConnectorAzureVnetDelete(ctx context.Context, d *schema.ResourceDat
 
 // generateConnectorAzureVnetRequest generate request for connector-azure-vnet
 func generateConnectorAzureVnetRequest(d *schema.ResourceData, m interface{}) (*alkira.ConnectorAzureVnet, error) {
-
 	//
 	// Segment
 	//
 	segmentName, err := getSegmentNameById(d.Get("segment_id").(string), m)
-
```
Comment (spikewang) [inline alkira/resource_alkira_connector_azure_vnet.go:411]: not sure why all those lines change, seems like not related. Please keep the change related to this PR, so it's easier to see what's the real change.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		CreateContext: resourceAlkiraPeeringGatewayCxpCreate,
+		ReadContext:   resourceAlkiraPeeringGatewayCxpRead,
+		UpdateContext: resourceAlkiraPeeringGatewayCxpUpdate,
+		DeleteContext: resourceAlkiraPeeringGatewayCxpDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "The name of the Peering Gateway.\nOnce deployed this property can not be changed. All the changes will be ignored.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:26]: Try to keep all those description at 80 per line... so it will be much easier to manage.

Comment (spikewang) [reply alkira/resource_alkira_peering_gateway_cxp.go:26]: And "all changes will be ignored" is not a good behaviour... we should fix that and API should report errors back if someone tries to update this field.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"description": {
+				Description: "Description of the Peering Gateway.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:31]: Make the naming consistent, `Description of the CXP Peering Gateway`, since there is other peering gateway resources.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"cxp": {
+				Description: "The CXP to which the Gateway is attached.\nOnce deployed this property can not be changed. All the changes will be ignored.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:36]: same here... we simple can't say that "all changes will be ignored"... API should return proper error...

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+
+			// TODO: change cloud_provider to be a Required value when more cloud providers are added and remove the default value.
+			"cloud_provider": {
+				Description: "The cloud provider on which the gateway will be created.\nOnce deployed this property can not be changed. All the changes will be ignored.",
+				Type:        schema.TypeString,
+				// Required:    true,
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:45]: Remove

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+				Optional: true,
+				Default:  "AZURE",
+			},
+			"cloud_region": {
+				Description: "The cloud region on which the ATH will be created. Eg : eastus , westus.\nOnce deployed this property can not be changed. All the changes will be ignored.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:50]: Rework this to be more clear:

The region of the specified cloud provider on which the resource should be created. E.g.  if `cloud_provider` is `AZURE`, the region could be `eastus`.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+			},
+			"cxp": {
+				Description: "The CXP to which the Gateway is attached.\nOnce deployed this property can not be changed. All the changes will be ignored.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+
+			// TODO: change cloud_provider to be a Required value when more cloud providers are added and remove the default value.
+			"cloud_provider": {
+				Description: "The cloud provider on which the gateway will be created.\nOnce deployed this property can not be changed. All the changes will be ignored.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:43]: I don't think any other providers are supported at this time, so say it clearly in Description:

The cloud provider on which the resource will be created. Default value is `AZURE` and only `AZURE` is supported for now.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+				Optional: true,
+				Default:  "AZURE",
+			},
+			"cloud_region": {
+				Description: "The cloud region on which the ATH will be created. Eg : eastus , westus.\nOnce deployed this property can not be changed. All the changes will be ignored.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"segment": {
+				Description: "The name of the segment in which the gateway is created.\nOnce deployed this property can not be changed. All the changes will be ignored.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:55]: We do use `segment_id` in TF. This should be consistent with all other resources. We will take `segment_id` and convert to segment name internally. Refer to other resources.

**Review by spikewang — CHANGES_REQUESTED** [review]

File: examples/resources/alkira_peering_gateway_cxp/resource.tf [hunk 12]
```diff
@@ -0,0 +1,7 @@
+resource "alkira_peering_gateway_cxp" "test1" {
+  name         = "tf-test-1"
```
Comment (spikewang) [inline examples/resources/alkira_peering_gateway_cxp/resource.tf:2]: For example, we try to move away from referring to our internal test naming. I am fixing all other examples. so we could just use name `example`. Tweak all `test` to `example`.

File: alkira/resource_alkira_connector_azure_vnet.go (hunk 1 above)
Comment (asim-ak) [reply alkira/resource_alkira_connector_azure_vnet.go:68]: sorry, this is caused by "gofumpt" formatting.

File: alkira/resource_alkira_peering_gateway_cxp.go (hunk 5 above)
Comment (asim-ak) [reply alkira/resource_alkira_peering_gateway_cxp.go:26]: updated

File: alkira/resource_alkira_peering_gateway_cxp.go (hunk 7 above)
Comment (asim-ak) [reply alkira/resource_alkira_peering_gateway_cxp.go:36]: We had a discussion during the ACG meeting, basically if a resource cannot be updated it would just ignore the changes, if the update was conditional , it would throw an error. 
So in this case the changes would be just ignored. I was asked to add this note in the documentation.

File: alkira/resource_alkira_peering_gateway_cxp.go (hunk 8 above)
Comment (asim-ak) [reply alkira/resource_alkira_peering_gateway_cxp.go:45]: done.

File: alkira/resource_alkira_peering_gateway_cxp.go (hunk 9 above)
Comment (asim-ak) [reply alkira/resource_alkira_peering_gateway_cxp.go]: done.

File: alkira/resource_alkira_peering_gateway_cxp.go (hunk 11 above)
Comment (asim-ak) [reply alkira/resource_alkira_peering_gateway_cxp.go]: done.

File: examples/resources/alkira_peering_gateway_cxp/resource.tf (hunk 12 above)
Comment (asim-ak) [reply examples/resources/alkira_peering_gateway_cxp/resource.tf]: done.

File: alkira/resource_alkira_peering_gateway_cxp.go (hunk 10 above)
Comment (asim-ak) [reply alkira/resource_alkira_peering_gateway_cxp.go]: done.

File: alkira/resource_alkira_connector_azure_vnet.go (hunk 1 above)
Comment (spikewang) [reply alkira/resource_alkira_connector_azure_vnet.go:68]: lol, we used the standard `gofmt`, so it should be okey. It's funny that the tool says that it won't generate diff from `gofmt`. Hehe.

File: alkira/resource_alkira_peering_gateway_cxp.go (hunk 7 above)
Comment (spikewang) [reply alkira/resource_alkira_peering_gateway_cxp.go:36]: I disagree. This doesn't belong to ACG team. It's just a bad user experience. If backend doesn't want to fix this, please do validation in `update` function.

Comment (spikewang) [reply alkira/resource_alkira_peering_gateway_cxp.go:36]: Even we wanted to have this description. It should be like this and have line break at 80 like this:

```
Description: "The CXP to which the Gateway is attached to. " +
                      "Once deployed, this field can't be updated.",
```

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+	d.Set("state", resource.State)
+
+	return nil
+}
+
+func resourceAlkiraPeeringGatewayCxpUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	// INIT
+	api := alkira.NewPeeringGatewayCxp(m.(*alkira.AlkiraClient))
+
+	request, err := generateAlkiraCxpPeeringGatewayRequest(d, m)
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:137]: If we don't actually support update, then let's error out here directly, instead of commenting on almost every single field, which doesn't make any sense.

**Review by spikewang — CHANGES_REQUESTED** [review]

File: alkira/resource_alkira_connector_azure_vnet.go (hunk 2 above)
Comment (spikewang) [reply alkira/resource_alkira_connector_azure_vnet.go:114]: Again, I am not really sure why benefit to have this new format. `gofmt` is built-in and generic enough across all our Golang services. We should stick with that.

**Review by spikewang — CHANGES_REQUESTED** [review]

**Comment by asim-ak** [comment]
@spikewang regarding 'gofumpt' my understanding is that it is much more stricter version of 'gofmt' , which leads it to be more consistent. 
I've reverted to to 'gofmt' for now.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+				Optional:    true,
+			},
+			"cxp": {
+				Description: "The CXP to which the Gateway is attached." +
+					"Once deployed, this property can not be changed.",
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			// TODO: change cloud_provider to be a Required value when more cloud providers are added and remove the default value.
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:42]: Remove this. This is not needed. We will have task tracking this if new provider is added later on.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+			"cloud_provider": {
+				Description: "The cloud provider on which the resource will be created." +
+					"Default value is AZURE and only AZURE is supported for now.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:45]: Add AZURE into markdown code by using "\`". It should be like \`azure\`.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "AZURE",
+			},
+			"cloud_region": {
+				Description: "The region of the specified cloud provider on which the resource should be created." +
+					"E.g. if cloud_provider is AZURE, the region could be eastus.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:52]: same here, cloud_provider should be in as code block. The line seems like still too long to me. 80 break.

"The region of the specified ``cloud_provider`` on which the resource " +
"should be created. For example, if ``cloud_provider`` is ``AZURE``, the " +
"region could be ``eastus``.",

Also, notice that at the line end, you will need to leave a space there.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"segment_id": {
+				Description: "The ID of the segment in which the gateway is created." +
+					"Once deployed, this property can not be changed.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:58]: Remove this property can not be changed.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+		ReadContext:   resourceAlkiraPeeringGatewayCxpRead,
+		UpdateContext: resourceAlkiraPeeringGatewayCxpUpdate,
+		DeleteContext: resourceAlkiraPeeringGatewayCxpDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "The name of the Peering Gateway." +
+					"Once deployed, this property can not be changed.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:26]: Remove this line. No need.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+				Required: true,
+			},
+			"description": {
+				Description: "Description of the CXP Peering Gateway.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"cxp": {
+				Description: "The CXP to which the Gateway is attached." +
+					"Once deployed, this property can not be changed.",
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:37]: Remove this line.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+	d.Set("cloud_region", resource.CloudRegion)
+	d.Set("segment_id", segmentId)
+	d.Set("state", resource.State)
+
+	return nil
+}
+
+func resourceAlkiraPeeringGatewayCxpUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	// checking if only description field is changed
+	if d.HasChange("description") && !d.HasChanges("name", "cloud_provider", "cloud_region", "segment_id") {
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:139]: Honestly, if the only field could be changed is description, we should just fail and return. It's pointless to just check description. Just return directly. Simple and clean.

**Review by spikewang — CHANGES_REQUESTED** [review]
I will show you how to use the template file.

File: alkira/resource_alkira_peering_gateway_cxp.go [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+	api := alkira.NewPeeringGatewayCxp(m.(*alkira.AlkiraClient))
+
+	request, err := generateAlkiraCxpPeeringGatewayRequest(d, m)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// UPDATE
+	_, err, _ = api.Update(d.Id(), request)
+
```
Comment (spikewang) [inline alkira/resource_alkira_peering_gateway_cxp.go:158]: We need to check this `err` before return or just simplify this update. We don't need to support update at all. Just disable update return directly.

**Review by spikewang — CHANGES_REQUESTED** [review]

**Review by spikewang — APPROVED** [review]
